test/model/message_list_test.dart

```diff
@@ -0,0 +1,241 @@
+      check(idx).equals(-1);
+
+      //Is the second
+      idx = messageList.findMessageWithId(4);
+      check(idx).equals(1);
+
+      //Between second and third
+      idx = messageList.findMessageWithId(5);
+      check(idx).equals(-1);
+
+      //Is the third
+      idx = messageList.findMessageWithId(6);
+      check(idx).equals(2);
+
+      //After the third
+      idx = messageList.findMessageWithId(7);
+      check(idx).equals(-1);
+
+      //Invalid IDs
+      idx = messageList.findMessageWithId(-8409);
+      check(idx).equals(-1);
+
+      idx = messageList.findMessageWithId(0);
+      check(idx).equals(-1);
+    });
+
+    test('update events are correctly applied to message when it is in the stream', () async {
+      final store = await setupStore(stream);
+
+      const oldContent = "<p>Hello, world</p>";
+      const newContent = "<p>Hello, edited</p>";
+      const newTimestamp = 99999;
+
+      List<String> oldFlags = [];
+      List<String> newFlags = ["starred"];
+
+      final mockMessage = eg.streamMessage(id: 243, stream: stream, content: oldContent, flags: oldFlags);
+      final messageList = await messageListViewWithMessages([mockMessage], store, narrow);
+
+      final updateEvent = UpdateMessageEvent(
+        id: 1,
+        messageId: mockMessage.id,
+        messageIds: [mockMessage.id],
+        flags: newFlags,
+        renderedContent: newContent,
+        editTimestamp: newTimestamp,
+        isMeMessage: true,
+        userId: userId
+      );
+
+      final message = messageList.messages[0];
+      check(message)
+        ..content.equals(oldContent)
+        ..flags.deepEquals(oldFlags)
+        ..isMeMessage.equals(false);
+
+      var listenersNotified = false;
+
+      messageList.addListener(() { listenersNotified = true; });
+      messageList.maybeUpdateMessage(updateEvent);
+
+      final updatedMessage = messageList.messages[0];
+      check(updatedMessage).identicalTo(message);
+      check(listenersNotified).equals(true);
+
+      check(message)
+        ..content.equals(newContent)
+        ..lastEditTimestamp.equals(newTimestamp)
+        ..flags.equals(newFlags)
+        ..isMeMessage.equals(true);
+    });
+
+    test('update event is ignored when message is not in the message list', () async {
+      final store = await setupStore(stream);
+
+      const oldContent = "<p>Hello, world</p>";
+      const newContent = "<p>Hello, edited</p>";
+      const newTimestamp = 99999;
+
+      final mockMessage = eg.streamMessage(id: 243, stream: stream, content: oldContent);
+      final messageList = await messageListViewWithMessages([mockMessage], store, narrow);
+
+      final updateEvent = UpdateMessageEvent(
+        id: 1,
+        messageId: 972,
+        messageIds: [972],
+        flags: mockMessage.flags,
+        renderedContent: newContent,
+        editTimestamp: newTimestamp,
+        userId: userId,
+      );
+
+      final message = messageList.messages[0];
```

Comment:
A handy alternative way to write this is with the [`Iterable.single`](https://api.dart.dev/stable/3.0.7/dart-core/Iterable/single.html) getter:
```suggestion
      final message = messageList.messages.single;
```

That's equivalent to `[0]` when there actually is exactly one element, but throws if there's (fewer or) more than one. As a result it makes it explicit to the reader that there's supposed to be just one element here (and in the context of a test, it lets the reader *know* there's just one element here as long as the test is passing), rather than thinking this might be the first of several.

In particular in this test which is checking that nothing happens, it lets the reader immediately stop worrying about whether some other message did get updated and the test is just looking at the wrong one.

Reply:
Convenient!